File: week2/quicksort/quicksort.rb
class Quicksort

  def sort(array)
    comparisons = 0
    unless array.nil? or array.empty?
      first = 0
      last = array.size - 1
      comparisons = quicksort(array, first, last, 0)
    end
    [array, comparisons]
  end

  private

  def quicksort(array, left, right, comparisons)
    if right > left
      split_index = partition(array, left, right)
      comparisons += (right - left)
      comparisons = quicksort(array, left, split_index - 1, comparisons)
      comparisons = quicksort(array, split_index + 1, right, comparisons)
    end
    comparisons
  end

  def partition(array, first, last)
    p = choose_index_of_pivot(array, first, last)
    swap(array, first, p) unless p == first

    i = j = first + 1
    while j <= last
      if array[j] < array[first]
        swap(array, i, j)
        i += 1
      end
      j += 1
    end

    pivot_place = i - 1
    swap(array, first, pivot_place)
    pivot_place
  end

  def choose_index_of_pivot(array, first, last)
    # choose_random_index(first, last)
    # choose_first_index(first, last)
    # choose_last_index(first, last)
    choose_median_of_three(array, first, last)
  end

  def choose_random_index(first, last)
    if last - first > 1
      first + Random.rand(last - first)
    else
      first
    end
  end

  def choose_first_index(first, last)
    first
  end

  def choose_last_index(first, last)
    last
  end

  def calculate_midpoint(first, last)
    size = last - first + 1
    if size % 2 == 0
      first + size/2 - 1
    else
      first + size/2
    end
  end

  def choose_median_of_three(array, first, last)
    if last - first > 1
      smallest = median = first
      midpoint = calculate_midpoint(first, last)

      if array[first] < array[midpoint]
        median = midpoint
      else
        smallest = midpoint
      end

      if array[last] < array[smallest]
        median = smallest
      elsif array[last] < array[median]
        median = last
      end

      median
    else
      first
    end
  end

  def swap(array, i, j)
    unless i == j
      array[i] += array[j]
      array[j] = array[i] - array[j]
      array[i] -= array[j]
    end
  end
end

File: week2/c-quicksort/quicksort.c
/*
 * =====================================================================================
 *
 *       Filename:  quicksort.c
 *
 *    Description:  
 *
 *        Version:  1.0
 *        Created:  23/10/14 22:14:22
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  Dragan Glumac (draganglumac), dragan.glumac@gmail.com
 *   Organization:  
 *
 * =====================================================================================
 */
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include "quicksort.h"

void kwiksort(int *array, int first, int last); 
int choose_pivot(int *array, int first, int last);
void swap(int *array, int i, int j);
int partition(int *array, int first, int last);

void quicksort(int *array, int size) {
	if (array == 0 || size <= 0) {
		return;
	}
	kwiksort(array, 0, size - 1);
}
void kwiksort(int *array, int first, int last) {
	if (first >= last) {
		return;
	}
	int pivot = partition(array, first, last);
	kwiksort(array, first, pivot - 1);
	kwiksort(array, pivot + 1, last);
}
int choose_first(int *array, int first, int last) {
	return first;
}
int choose_last(int *array, int first, int last) {
	return last;
}
int calculate_midpoint(int first, int last) {
	int size = last - first + 1;
	if (size % 2 == 0) {
		return first + size/2 - 1;
	}
	else {
		return first + size/2;
	}
}
int choose_median_of_three(int *array, int first, int last) {
	if (last - first > 1) {
		int smallest = first, median = first;
		int midpoint = calculate_midpoint(first, last);
		if (array[first] < array[midpoint]) {
			median = midpoint;
		}
		else {
			smallest = midpoint;
		}
		if (array[last] < array[smallest]) {
			median = smallest;
		}
		else if (array[last] < array[median]) {
			median = last;
		}
		return median;
	}
	else {
		return first;
	}
}
int choose_random(int *array, int first, int last) {
	int pivot = first;
	if (last - first > 1) {
		srand(time(0));	
		pivot = first + rand() % (last - first);
	}
	return pivot;
}
int choose_pivot(int *array, int first, int last) {
//	return choose_first(array, first, last);
//	return choose_last(array, first, last);
//	return choose_median_of_three(array, first, last);
	return choose_random(array, first, last);
}
void swap(int *array, int i, int j) {
	if (i != j) {
		array[i] += array[j];
		array[j] = array[i] - array[j];
		array[i] -= array[j];
	}
}
int partition(int *array, int first, int last) {
	int pivot = choose_pivot(array, first, last);
	if (pivot != first) {
		swap(array, first, pivot);
	}
	int i = first + 1, j = first + 1;
	while (j <= last) {
		if (array[j] < array[first]) {
			swap(array, i, j);
			i++;
		}
		j++;
	}
	pivot = i - 1;
	swap(array, first, pivot);
	return pivot;
}
